diff -ru dvsdk/dmai_2_20_00_15/packages/ti/sdo/dmai/linux/omap3530/Capture.c dvsdk.new/dmai_2_20_00_15/packages/ti/sdo/dmai/linux/omap3530/Capture.c
--- dvsdk/dmai_2_20_00_15/packages/ti/sdo/dmai/linux/omap3530/Capture.c	2013-08-14 12:54:51.369671248 -0700
+++ dvsdk.new/dmai_2_20_00_15/packages/ti/sdo/dmai/linux/omap3530/Capture.c	2013-08-14 12:55:24.909663631 -0700
@@ -193,7 +193,7 @@
 	index = 0;
 	do {
 		memset(&media->entity[index], 0, sizeof(struct media_entity_desc));
-		media->entity[index].id = index | MEDIA_ENTITY_ID_FLAG_NEXT;
+		media->entity[index].id = index | MEDIA_ENT_ID_FLAG_NEXT;
 		ret = ioctl(media->media_fd, MEDIA_IOC_ENUM_ENTITIES, &media->entity[index]);
 		if (ret < 0) {
 			break;
@@ -247,12 +247,12 @@
 			for(i = 0; i < media->entity[index].links; i++) {
 				link.source.entity = links.links->source.entity;
 				link.source.index = links.links->source.index;
-				link.source.flags = MEDIA_PAD_FLAG_OUTPUT;
+				link.source.flags = MEDIA_PAD_FL_SOURCE;
 				link.sink.entity = links.links->sink.entity;
 				link.sink.index = links.links->sink.index;
-				link.sink.flags = MEDIA_PAD_FLAG_INPUT;
-				link.flags = (link.flags & ~MEDIA_LINK_FLAG_ENABLED) |
-					(link.flags & MEDIA_LINK_FLAG_IMMUTABLE);
+				link.sink.flags = MEDIA_PAD_FL_SINK;
+				link.flags = (link.flags & ~MEDIA_LNK_FL_ENABLED) |
+					(link.flags & MEDIA_LNK_FL_IMMUTABLE);
 				ret = ioctl(media->media_fd, MEDIA_IOC_SETUP_LINK, &link);
 				if(ret)
 					break;
@@ -284,7 +284,7 @@
 				printf("pads for entity %d=", media->entity[index].id);
 			for(i = 0 ; i < media->entity[index].pads; i++) {
 				printf("(%d %s) ", links.pads->index,
-						(links.pads->flags & MEDIA_PAD_FLAG_INPUT) ?
+						(links.pads->flags & MEDIA_PAD_FL_SINK) ?
 						"INPUT" : "OUTPUT");
 				links.pads++;
 			}
@@ -295,7 +295,7 @@
 						links.links->source.index,
 						links.links->sink.entity,
 						links.links->sink.index);
-				if(links.links->flags & MEDIA_LINK_FLAG_ENABLED)
+				if(links.links->flags & MEDIA_LNK_FL_ENABLED)
 					printf("\tACTIVE\n");
 				else
 					printf("\tINACTIVE \n");
@@ -313,14 +313,14 @@
 	printf("Enabling link [tvp5146]===>[ccdc]\n");
 	memset(&link, 0, sizeof(link));
 
-	link.flags |=  MEDIA_LINK_FLAG_ENABLED;
+	link.flags |=  MEDIA_LNK_FL_ENABLED;
 	link.source.entity = input;
 	link.source.index = 0;	/* Only 1 pad */
-	link.source.flags = MEDIA_PAD_FLAG_OUTPUT;
+	link.source.flags = MEDIA_PAD_FL_SOURCE;
 
 	link.sink.entity = media->ccdc;
 	link.sink.index = 0; /* Sink pad of CCDC, 0 */
-	link.sink.flags = MEDIA_PAD_FLAG_INPUT;
+	link.sink.flags = MEDIA_PAD_FL_SINK;
 
 	ret = ioctl(media->media_fd, MEDIA_IOC_SETUP_LINK, &link);
 	if(ret) {
@@ -333,14 +333,14 @@
 	printf("Enabling link [ccdc]===>[video_node]\n");
 	memset(&link, 0, sizeof(link));
 
-	link.flags |=  MEDIA_LINK_FLAG_ENABLED;
+	link.flags |=  MEDIA_LNK_FL_ENABLED;
 	link.source.entity = media->ccdc;
 	link.source.index = 1; /* Source pad of CCDC: 1 */
-	link.source.flags = MEDIA_PAD_FLAG_OUTPUT;
+	link.source.flags = MEDIA_PAD_FL_SOURCE;
 
 	link.sink.entity = media->video;
 	link.sink.index = 0;
-	link.sink.flags = MEDIA_PAD_FLAG_INPUT;
+	link.sink.flags = MEDIA_PAD_FL_SINK;
 
 	ret = ioctl(media->media_fd, MEDIA_IOC_SETUP_LINK, &link);
 	if(ret)
diff -ru dvsdk/Makefile dvsdk.new/Makefile
--- dvsdk/Makefile	2011-12-22 11:00:00.000000000 -0800
+++ dvsdk.new/Makefile	2013-08-14 12:45:48.853426974 -0700
@@ -21,22 +21,22 @@
 #==============================================================================
 # Build target build necessary components to enable all other build targets.
 #==============================================================================
-components: linux cmem sdma dsplink lpm c6run c6accel codecs dmai  
+components: cmem sdma dsplink lpm c6run c6accel codecs dmai  
 
 #==============================================================================
 # Clean up the targets built by 'make all'.
 #==============================================================================
-clean:	linux_clean dsplink_clean cmem_clean lpm_clean sdma_clean c6run_clean codecs_clean dmai_clean c6accel_clean ceexamples_clean u-boot_clean  demos_clean psp_examples_clean opengldemos_clean xload_clean dvtb_clean
+clean:	dsplink_clean cmem_clean lpm_clean sdma_clean c6run_clean codecs_clean dmai_clean c6accel_clean
 
 #==============================================================================
 # Build everything rebuildable.
 #==============================================================================
-all: components u-boot ceexamples demos psp_examples opengldemos xload c6run_examples matrix gstreamer_ti dvtb
+all: components gstreamer_ti
 
 #==============================================================================
 # Install everything
 #==============================================================================
-install: linux_install cmem_install sdma_install opengldemos_install ceexamples_install codecs_install lpm_install dsplink_install u-boot_install dmai_install demos_install  psp_examples_install c6accel_install xload_install c6run_install matrix_install gstreamer_ti_install loadmodule_install dvtb_install
+install: cmem_install sdma_install codecs_install lpm_install dsplink_install dmai_install c6accel_install c6run_install gstreamer_ti_install loadmodule_install
 
 #==============================================================================
 # A help message target.
@@ -806,7 +806,7 @@
 	LPM_INSTALL_DIR="$(LPM_INSTALL_DIR)" \
 	ARM_TOOLCHAIN_PATH="$(CSTOOL_DIR)" \
 	ARM_TOOLCHAIN_PREFIX="arm-arago-linux-gnueabi-" \
-	DSP_REGION_BASE_ADDR=0x83700000 \
+	DSP_REGION_BASE_ADDR=0x90000000 \
 	DSP_REGION_CMEM_SIZE=0x02200000 \
 	DSP_REGION_CODE_SIZE=0x02700000 \
 	dm3730_config
diff -ru dvsdk/Rules.make dvsdk.new/Rules.make
--- dvsdk/Rules.make	2013-08-13 13:59:21.872392933 -0700
+++ dvsdk.new/Rules.make	2013-08-14 13:12:09.053696497 -0700
@@ -55,7 +55,7 @@
 PSP_INSTALL_DIR=$(DVSDK_INSTALL_DIR)/psp
 
 # The directory that points to your kernel source directory.
-LINUXKERNEL_INSTALL_DIR=$(DVSDK_INSTALL_DIR)/psp/linux-2.6.37-psp04.02.00.07.sdk
+#LINUXKERNEL_INSTALL_DIR=$(DVSDK_INSTALL_DIR)/psp/linux-2.6.37-psp04.02.00.07.sdk
 KERNEL_INSTALL_DIR=$(LINUXKERNEL_INSTALL_DIR)
 
 # Where the local power manager is installed.
@@ -74,7 +74,7 @@
 GRAPHICS_INSTALL_DIR=$(DVSDK_INSTALL_DIR)/graphics-sdk_4.03.00.02
 
 # The prefix to be added before the GNU compiler tools (optionally including # path), i.e. "arm_v5t_le-" or "/opt/bin/arm_v5t_le-".
-CSTOOL_DIR=/home/garrett/dvsdk/linux-devkit/
+CSTOOL_DIR=$(DVSDK_INSTALL_DIR)/linux-devkit/
 CSTOOL_PREFIX=$(CSTOOL_DIR)/bin/arm-arago-linux-gnueabi-
 CSTOOL_PATH=$(CSTOOL_DIR)/bin
 
@@ -82,5 +82,5 @@
 MVTOOL_PREFIX=$(CSTOOL_PREFIX)
 
 # Where to copy the resulting executables
-EXEC_DIR=$(HOME)/install/$(PLATFORM)
+#EXEC_DIR=$(HOME)/install/$(PLATFORM)
 
