diff -ru git/cmake/OpenCVDetectPython.cmake git.new/cmake/OpenCVDetectPython.cmake
--- git/cmake/OpenCVDetectPython.cmake	2013-08-21 10:39:37.000000000 -0700
+++ git.new/cmake/OpenCVDetectPython.cmake	2013-08-21 11:03:16.777003086 -0700
@@ -81,10 +81,14 @@
 
     if(NOT PYTHON_NUMPY_INCLUDE_DIR)
       # Attempt to discover the NumPy include directory. If this succeeds, then build python API with NumPy
-      execute_process(COMMAND ${PYTHON_EXECUTABLE} -c "import os; os.environ['DISTUTILS_USE_SDK']='1'; import numpy.distutils; print numpy.distutils.misc_util.get_numpy_include_dirs()[0]"
+      if(NOT DEFINED PYTHON_NUMPY_INCLUDE_DIRS)
+          execute_process(COMMAND ${PYTHON_EXECUTABLE} -c "import os; os.environ['DISTUTILS_USE_SDK']='1'; import numpy.distutils; print numpy.distutils.misc_util.get_numpy_include_dirs()[0]"
                       RESULT_VARIABLE PYTHON_NUMPY_PROCESS
-                      OUTPUT_VARIABLE PYTHON_NUMPY_INCLUDE_DIR
+                      OUTPUT_VARIABLE PYTHON_NUMPY_INCLUDE_DIRS
                       OUTPUT_STRIP_TRAILING_WHITESPACE)
+      else()
+          set(PYTHON_NUMPY_PROCESS 0)
+      endif()
 
       if(PYTHON_NUMPY_PROCESS EQUAL 0)
         file(TO_CMAKE_PATH "${PYTHON_NUMPY_INCLUDE_DIR}" _PYTHON_NUMPY_INCLUDE_DIR)
