diff -ru gst-plugins-good-0.10.31/sys/v4l2/gstv4l2object.c gst-plugins-good-0.10.31.new/sys/v4l2/gstv4l2object.c
--- gst-plugins-good-0.10.31/sys/v4l2/gstv4l2object.c	2011-12-30 05:59:13.000000000 -0800
+++ gst-plugins-good-0.10.31.new/sys/v4l2/gstv4l2object.c	2013-08-23 13:27:38.610288336 -0700
@@ -1562,8 +1562,8 @@
   gint int_height = height;
 
   /* interlaced detection using VIDIOC_TRY/S_FMT */
-  if (!gst_v4l2_object_get_nearest_size (v4l2object, pixelformat,
-          &int_width, &int_height, &interlaced))
+//  if (!gst_v4l2_object_get_nearest_size (v4l2object, pixelformat,
+//          &int_width, &int_height, &interlaced))
     return NULL;
 
   memset (&ival, 0, sizeof (struct v4l2_frmivalenum));
