diff -ru OpenCV-2.4.3/modules/highgui/src/cap_v4l.cpp OpenCV-2.4.3.new/modules/highgui/src/cap_v4l.cpp
--- OpenCV-2.4.3/modules/highgui/src/cap_v4l.cpp	2012-10-17 15:55:38.000000000 -0700
+++ OpenCV-2.4.3.new/modules/highgui/src/cap_v4l.cpp	2013-09-13 08:32:36.395209868 -0700
@@ -529,7 +529,11 @@
 
 static int autosetup_capture_mode_v4l2(CvCaptureCAM_V4L* capture)
 {
-  if (try_palette_v4l2(capture, V4L2_PIX_FMT_BGR24) == 0)
+  if (try_palette_v4l2(capture, V4L2_PIX_FMT_UYVY) == 0)
+  {
+    capture->palette = PALETTE_UYVY;
+  }
+  else if (try_palette_v4l2(capture, V4L2_PIX_FMT_BGR24) == 0)
   {
     capture->palette = PALETTE_BGR24;
   }
@@ -563,10 +567,6 @@
   {
     capture->palette = PALETTE_YUYV;
   }
-  else if (try_palette_v4l2(capture, V4L2_PIX_FMT_UYVY) == 0)
-  {
-    capture->palette = PALETTE_UYVY;
-  }
   else
   if (try_palette_v4l2(capture, V4L2_PIX_FMT_SN9C10X) == 0)
   {
